src/ApplicationInsights.Kubernetes/K8sInfoBackgroundService.cs

```diff
@@ -0,0 +1,90 @@
+        IOptions<AppInsightsForKubernetesOptions> options)
+    {
+        _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+        _k8SEnvironmentHolder = k8SEnvironmentHolder ?? throw new ArgumentNullException(nameof(k8SEnvironmentHolder));
+        _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+    }
+
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        while (true)
+        {
+            try
+            {
+                _logger.LogDebug($"Starting {nameof(UpdateK8sEnvironmentAsync)}");
+                await UpdateK8sEnvironmentAsync(stoppingToken).ConfigureAwait(false);
+                TimeSpan interval = _options.ClusterInfoRefreshInterval;
+                _logger.LogDebug($"Finished {nameof(UpdateK8sEnvironmentAsync)}, delay for {interval}");
+                await Task.Delay(interval);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception ex)
+            {
+                // In case of any unexpected exception, we shall log it than let it throw.
+                _logger.LogError("Unexpected error happened. Telemetry enhancement with Kubernetes info won't happen. Message: {0}", ex.Message);
+                _logger.LogTrace("Unexpected error happened. Telemetry enhancement with Kubernetes info won't happen. Details: {0}", ex.ToString());
+            }
+#pragma warning restore CA1031 // Do not catch general exception types
+        }
+    }
+
+    private async Task UpdateK8sEnvironmentAsync(CancellationToken cancellationToken)
+    {
+        await using (AsyncServiceScope scope = _serviceProvider.CreateAsyncScope())
+        {
+            // Get scoped service ready
+            IServiceProvider provider = scope.ServiceProvider;
+            IK8sEnvironmentFactory factory = provider.GetRequiredService<IK8sEnvironmentFactory>();
+
+            // Prepare cancellation token with timeout.
+            using CancellationTokenSource timeoutSource = new CancellationTokenSource(_options.InitializationTimeout);
+            CancellationToken timeoutToken = timeoutSource.Token;
+            
+            try
+            {
+                using (CancellationTokenSource linkedTokeSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutToken))
+                {
+
+                    IK8sEnvironment? environment = await factory.CreateAsync(cancellationToken: linkedTokeSource.Token).ConfigureAwait(false);
+                    _k8SEnvironmentHolder.K8sEnvironment = environment;
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                if (timeoutToken.IsCancellationRequested)
+                {
+                    _logger.LogWarning("Querying Kubernetes cluster info timed out.");
+                    _k8SEnvironmentHolder.K8sEnvironment = null;
```

Comment:
This is probably fine because `IsQueryTimeout` will only be set once, and `K8sEnvironment` will never be anything other than `null` if `IsQueryTimeout == true`.

In general though, it gets pretty tricky when one thread is setting a few related properties, and other threads need to read all values in a consistent way. The only way to ensure correctness in such (general) case is to use locks or other concurrency mechanisms.

Reply:
Got rid of IsQueryTimeout already. I want something very very simple. It is too easy to make it more complex than needed.

Reply:
I was expecting that question :-) I want to discuss it. That is actually something I considered and here are my thoughts:

In general, I want to avoid complexity, so, I designed it a bit:

1. The K8sEnvHolder is a singleton, and it holds a ref to K8sEnv object - properties on K8sEnv doesn't mutate;
2. When we gather new info for K8sEnv, it builds a new K8sEnv object, and when it is ready, update the ref - [code](https://github.com/xiaomi7732/ApplicationInsights-Kubernetes/blob/72043d4cf9ea3ec58a4ccdea20ea43aa62c55cc7/src/ApplicationInsights.Kubernetes/K8sInfoBackgroundService.cs#L72);
3. On the consuming side (initializer), the first thing we do is to reference the current K8sEnv object and always use it - [code](https://github.com/xiaomi7732/ApplicationInsights-Kubernetes/blob/72043d4cf9ea3ec58a4ccdea20ea43aa62c55cc7/src/ApplicationInsights.Kubernetes/TelemetryInitializers/KubernetesTelemetryInitializer.cs#L37).

In general, I designed it that way to avoid unnecessary semaphores/locks. Do you think that would work? Or do you see anything that is missing?

Reply:
Yes, I think it will work. If you want to be 100% safe with lock-free code like this, make the reference to K8sEnvironment within the K8sEnvironmentHolder a volatile field.

Reply:
(unresolving to make sure you saw the previous comment, feel free to re-resolve)

Reply:
volatile it is.